gulpfile.coffee

```diff
@@ -1,69 +1,239 @@
-gulp = require('gulp')
-$ = require('gulp-load-plugins')()
+fs = require 'fs'
+gulp = require('gulp-help') require 'gulp'
+
 main_bower_files = require 'main-bower-files'
 del = require 'del'
 exec = require('child_process').exec
 minimist = require 'minimist'
 
+$ = do require 'gulp-load-plugins'
+
+
+dir_bower_components = 'bower_components'
+dir_main = 'main'
+dir_node_modules = 'node_modules'
+
+dir_lib = "#{dir_main}/lib"
+dir_libx = "#{dir_main}/libx"
+dir_temp = 'temp'
+dir_venv = "#{dir_temp}/venv"
+
+file_lib = "#{dir_lib}.zip"
+
+dir_static = "#{dir_main}/static"
+dir_src = "#{dir_static}/src"
+dir_script = "#{dir_src}/script"
+dir_style = "#{dir_src}/style"
+dir_ext = "#{dir_static}/ext"
+dir_min = "#{dir_static}/min"
+dir_storage = "#{dir_temp}/storage"
+
 
-root_dir = './main'
-static_dir = "#{root_dir}/static"
 paths =
+  ext: [
+      "#{dir_ext}/jquery/dist/jquery.js"
+      "#{dir_ext}/moment/moment.js"
+      "#{dir_ext}/nprogress/nprogress.js"
+      "#{dir_ext}/bootstrap-sweetalert/lib/sweet-alert.js"
+      "#{dir_ext}/bootstrap/js/alert.js"
+      "#{dir_ext}/bootstrap/js/button.js"
+      "#{dir_ext}/bootstrap/js/transition.js"
+      "#{dir_ext}/bootstrap/js/collapse.js"
+      "#{dir_ext}/bootstrap/js/dropdown.js"
+      "#{dir_ext}/bootstrap/js/tooltip.js"
+    ]
   clean: [
-      "#{static_dir}/dst"
-      "#{static_dir}/ext"
-      "#{static_dir}/min"
+      dir_min
+      './**/*.pyc'
+      './**/*.pyo'
+      './**/*.~'
+    ]
+  clean_all: [
+      dir_min
+      dir_bower_components
+      dir_node_modules
+      "#{dir_temp}/pip.guard"
+    ]
+  clean_py: [
+      dir_lib
+      dir_venv
+    ]
+  style: [
+      "#{dir_style}/style.less"
+    ]
+  script: [
+      "#{dir_script}/**/*.coffee"
     ]
   watch: [
-      "#{static_dir}/**/*.css"
-      "#{static_dir}/**/*.js"
-      "#{root_dir}/**/*.html"
-      "#{root_dir}/**/*.py"
+      "#{dir_static}/**/*.css"
+      "#{dir_static}/**/*.js"
+      "#{dir_main}/**/*.html"
+      "#{dir_main}/**/*.py"
     ]
 
 
-run = (option) ->
-  proc = exec "python -u run.py -#{option}"
-  proc.stderr.on 'data', (data) -> process.stderr.write data
-  proc.stdout.on 'data', (data) -> process.stdout.write data
+gulp.task 'script', false, ->
+  gulp.src paths.script
+    .pipe $.coffee().on 'error', $.util.log
+    .pipe $.concat 'script.js'
+    .pipe do $.uglify
+    .pipe gulp.dest "#{dir_min}/script"
+
+
+gulp.task 'script_dev', false, ->
+  gulp.src paths.script
+    .pipe do $.sourcemaps.init
+    .pipe $.coffee().on 'error', $.util.log
+    .pipe $.concat 'script.dev.js'
```

Comment:
The problem with this approach I faced in the past is as the JavaScript codebase grows the build starts to slow down due to the concat. I have used [gulp-inject](https://github.com/topless/gae-init/blob/angular/gulpfile.coffee), be aware that they have to be seperate async tasks in order to do all the injections (libs and scripts). There is also another plugin I think called gulp-newer which rebuilds only the changed file. Great work by the way +1!